File: week_1/credit/credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>

int get_first_n_digits(long long card_number, int n)
{
    {
        while (card_number >= pow(10, n))
            card_number /= 10;
    }
    return card_number;
}

int main(void)
{
    long long card_number = get_long("Number: ");
    int sum = 0;
    int digits = 0;
    int first_two_digits = get_first_n_digits(card_number, 2);

    if (card_number < 1e12)
    {
        printf("INVALID\n");
        return 0;
    }

    while (card_number > 0)
    {
        int last_digit = card_number % 10;

        if (digits % 2 == 0)
        {
            sum += last_digit;
        }
        else
        {
            int doubled = last_digit * 2;
            sum += doubled / 10 + doubled % 10;
        }

        card_number /= 10;
        digits++;
    }

    if (sum % 10 == 0)
    {
        if ((first_two_digits == 34 || first_two_digits == 37) && digits == 15)
        {
            printf("AMEX\n");
        }
        else if ((first_two_digits >= 51 && first_two_digits <= 55) && digits == 16)
        {
            printf("MASTERCARD\n");
        }
        else if ((first_two_digits / 10 == 4) && (digits == 13 || digits == 16))
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }

    return 0;
}

File: week_6/sentimental-credit/credit.py
from cs50 import get_string
from math import floor

VISA_PREFIX = "4"
AMEX_PREFIXES = ["34", "37"]
MASTERCARD_PREFIXES = ["51", "52", "53", "54", "55"]
INVALID_BRAND = "INVALID"


def clean_input(card_number):
    return card_number.strip()


def get_card_brand(card_number):
    if card_number.startswith(VISA_PREFIX) and len(card_number) == 13:
        return "VISA"
    elif card_number[:2] in AMEX_PREFIXES and len(card_number) == 15:
        return "AMEX"
    elif card_number[:2] in MASTERCARD_PREFIXES and len(card_number) == 16:
        return "MASTERCARD"
    elif card_number.startswith(VISA_PREFIX) and len(card_number) == 16:
        return "VISA"
    else:
        return INVALID_BRAND


def is_valid_card(card_number):
    card_number = int(card_number)
    last = 0
    temp = 0
    sec_to_last = 0

    while card_number > 0:
        last += card_number % 10
        card_number //= 10

        temp = (card_number % 10) * 2

        if temp >= 10:
            sec_to_last += temp % 10
            temp //= 10
            sec_to_last += floor(temp)
        else:
            sec_to_last += temp

        card_number //= 10

    return (last + sec_to_last) % 10 == 0


def main():
    card_number = get_string("Number: ")
    card_number = clean_input(card_number)
    brand = get_card_brand(card_number)

    if brand != INVALID_BRAND:
        if is_valid_card(card_number):
            print(brand)
        else:
            print(INVALID_BRAND)
    else:
        print(INVALID_BRAND)


if __name__ == "__main__":
    main()
